rest-server/admin/src/main/scala/com/actionml/admin/MongoAdministrator.scala

```diff
@@ -17,53 +17,77 @@
-import com.mongodb.casbah.Imports._
-import com.mongodb.casbah.commons.MongoDBObject
+import org.mongodb.scala.MongoCollection
+import org.mongodb.scala.bson.ObjectId
+import org.mongodb.scala.bson.collection.immutable.Document
+import scaldi.{Injectable, Injector, Module}
 
-class MongoAdministrator extends Administrator with JsonParser with Mongo {
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success, Try}
 
-  lazy val enginesCollection: MongoCollection = connection("harness_meta_store")("engines")
-  lazy val commandsCollection: MongoCollection = connection("harness_meta_store")("commands") // async persistent though temporary commands
+class MongoAdministrator(override implicit val injector: Module)
+  extends Administrator with JsonParser with Mongo with Injectable {
+
+  lazy val enginesCollection: MongoCollection[CommonEngineData] = getDatabase("harness_meta_store").getCollection("engines")
+//  lazy val commandsCollection: MongoCollection[Engine] = getDatabase("harness_meta_store").getCollection("commands") // async persistent though temporary commands
   var engines = Map.empty[EngineId, Engine]
+  private val ec = inject[ExecutionContext]
 
   drawActionML
   private def newEngineInstance(engineFactory: String): Engine = {
-    Class.forName(engineFactory).newInstance().asInstanceOf[Engine]
+    //Class.forName(engineFactory).newInstance().asInstanceOf[Engine]
+    //val c = Class.getDeclaredConstructor(String.class).newInstance("HERESMYARG")
+    val v = Class.forName(engineFactory).getConstructors
+    val c = Injector.getClass
+    val c2 = classOf[Injector]
+    Class.forName(engineFactory).getDeclaredConstructor(classOf[Injector]).newInstance(injector).asInstanceOf[Engine]
   }
 
   // instantiates all stored engine instances with restored state
-  override def init() = {
+  override def init(): MongoAdministrator = {
+    implicit val _ = ec
+    def initEngine(engine: CommonEngineData): Future[Try[Engine]] = {
+      newEngineInstance(engine.engineFactory)
+        .initAndGet(engine.params)
+        .map(Success(_))
+        .recoverWith { case error =>
+          // it is possible that previously valid metadata is now bad, the Engine code must have changed
+          logger.error(s"Error creating engineId: ${engine._id} from ${engine.params}" +
+            s"\n\nTry deleting the engine instance by hand and re-adding to start fresh or fix the JSON config if there " +
+            s"is an error in it. Note: " +
+            s"this may happen when code for one version of the Engine has chosen to not be backwards compatible.")
+          // Todo: we need a way to cleanup in this situation
+          enginesCollection.deleteOne(Document("engineId" -> engine._id)).toFuture().map(_ => Failure(error))
+          // can't do this because the instance is null: deadEngine.destroy(), maybe we need a compan ion object with a cleanup function?
```

Comment:
don't delete it from the DB. I have changed the error message. Do nothing but report the error and return an error code to the REST API.

Also `${engine._id}` should be the engineId. You should re-define the case class you use for the EngineInstance data in Mongo to have `_id: String` and use the engineId as the _id key.